### PR TITLE
1710 various problems in access control

### PR DESCRIPTION
**Pull request by alvacouch**
1. via_group and via_user arguments added to get_resources_with_effective_access. 
2. excess antibugging removed. 

See issue #1710

### CODE REVIEW

**Comment by alvacouch** [comment]
There is one catch to enabling both `via_user` and `via_group` parameters: 

**If you enable both, either can squash the other,** in the sense that we always eliminate things with higher privilege from any enabled source. Example:

* `Garfield` shares group `Cats` with `Sylvester` as `CHANGE`
* `Garfield` shares resource `CatFood` with `Cats` as `CHANGE`
* (Now `Sylvester` has `CHANGE` over `CatFood`)
* `Garfield` shares resource `CatFood` with `Sylvester` as `OWNER`

Then
```Sylvester.uaccess.get_resources_with_explicit_access(PrivilegeCodes.CHANGE, via_user=True, via_group=True)```
does not contain `CatFood`, because `Sylvester` is an `OWNER` through **user** privilege, which "squashes" the group privilege, being more powerful.

via_user=True, via_group=False is the default. Thus this is plug compatible with the existing UI.

**Comment by alvacouch** [comment]
@pkdash I responded to @dtarb's feedback in #1622 by making the user privilege state machine independent of group privileges held at the time. 

This should not break anything, but might not change anything either, because the UI might not become aware that this is possible. 

For example, one can now set an **individual** privilege of VIEW or CHANGE over a resource for which one already holds a **group** privilege of CHANGE. Affected routines are `{can\_}?share_resource_with_user`.

**Comment by alvacouch** [comment]
@pkdash Done with post AGU updates. Ready for review.

File: hs_access_control/models.py [hunk 1]
```diff
@@ -1230,32 +1247,82 @@ def get_resources_with_explicit_access(self, this_privilege):
...
-                .filter(r2urp__privilege=this_privilege,
-                        r2urp__user=self.user)\
-                .exclude(pk__in=BaseResource.objects
-                         .filter(r2urp__user=self.user,
-                                 r2urp__privilege__lt=this_privilege))
+            if via_user:
+                return BaseResource.objects\
+                    .filter(r2urp__privilege=this_privilege,
+                            r2urp__user=self.user)\
+                    .exclude(pk__in=BaseResource.objects
```
Comment (pkdash) [inline hs_access_control/models.py]: Not sure what this would be excluding.

File: hs_access_control/models.py [hunk 2]
```diff
@@ -1230,32 +1247,82 @@ def get_resources_with_explicit_access(self, this_privilege):
...
+            else:
+                # groups can't own resources
+                return BaseResource.objects.empty()
 
         elif this_privilege == PrivilegeCodes.CHANGE:
             # CHANGE does not include immutable resources
+            uquery = Q(raccess__immutable=False,
+                       r2urp__privilege=this_privilege,
+                       r2urp__user=self.user)
+            uexcl = Q(raccess__immutable=False,
```
Comment (pkdash) [inline hs_access_control/models.py]: Same comment as above.

**Comment by pkdash** [comment]
@alvacouch I have now finished reviewing this. Have made few comments.

File: hs_access_control/models.py (hunk 1 above)
Comment (alvacouch) [inline hs_access_control/models.py]: Good point. Will remove exclusion.

**Comment by alvacouch** [comment]
@pkdash Responded to review comments and made changes.

**Comment by pkdash** [comment]
@alvacouch See if my comment for line#1266 in models.py makes any sense. Let  me know then I can +1.

**Comment by alvacouch** [comment]
Eliminated redundant filters. Found one more bug in user/group interplay. Added tests to ensure that bug does not recur. 

Question: `Resource.raccess.get_users_with_explicit_access` does not account for immutable flag. Should it?

**Comment by alvacouch** [comment]
@pkdash I think you might want to review the code for get_resources_with_explicit_access again. It is complex. Unfortunately, I think it needs to be.

File: hs_access_control/models.py [hunk 3]
```diff
@@ -1273,62 +1274,97 @@ def get_resources_with_explicit_access(self, this_privilege, via_user=True, via_
...
+                view = BaseResource.objects\
+                    .filter(query)
+
+            # step 2: compute immutable privileges, whose privilege is CHANGE
+            # with no other privilege < CHANGE (OWNER). Since groups can't own,
+            # there is no exclusion for them.
+            uquery = Q(raccess__immutable=True,
+                       r2urp__privilege=PrivilegeCodes.CHANGE,
+                       r2urp__user=self.user)
+            uexcl = Q(raccess__immutable=True,
```
Comment (pkdash) [inline hs_access_control/models.py]: I think this exclusion query is not needed.

File: hs_access_control/models.py [hunk 4]
```diff
@@ -1273,62 +1274,97 @@ def get_resources_with_explicit_access(self, this_privilege, via_user=True, via_
...
-        else:  # this_privilege == PrivilegeCodes.ViEW
+        else:  # this_privilege == PrivilegeCodes.VIEW
             # VIEW includes CHANGE+immutable as well as explicit VIEW
             # TODO: make this query more efficient, if possible!
             # CHANGE does not include immutable resources
+
+            # step one: compute view resources explicitly:
             uquery = Q(r2urp__privilege=this_privilege,
                        r2urp__user=self.user)
             uexcl = Q(r2urp__privilege__lt=this_privilege,
```
Comment (alvacouch) [inline hs_access_control/models.py]: See line 1314. This is the line that allows user privilege to override group privilege when both are looked at together. For example, suppose someone has VIEW from a group, and OWNER as a user. This line excludes owned resources from the VIEW resources.

File: hs_access_control/models.py (hunk 3 above)
Comment (alvacouch) [inline hs_access_control/models.py]: Note that calling this with `via_user=False`, `via_group=True` will not override via ownership, but will report only on group privileges. The exclusion is only invoked when one asks for a combined privilege report.

Comment (alvacouch) [inline hs_access_control/models.py]: Overall, the scheme here is that if one fixes the state of `via_user` and `via_group`, and then calls the routine for all three privilege codes, one should get a disjoint set for each call, and their union should be all resources.

Comment (pkdash) [inline hs_access_control/models.py]: The ```uquery``` is already filtering out  resources that are owned by the user. So I thought we don't need ```uexcl``` query. Let me know I can +1 this PR.

Comment (alvacouch) [inline hs_access_control/models.py]: I revised the logic to show how uexcl and gexcl are used.

**Comment by alvacouch** [comment]
@pkdash I rewrote the logic to clarify it. It should be much more clear what I am doing. See what you think.

**Comment by pkdash** [comment]
@alvacouch The changed query looks much better. It is a complex query and I assume it works as expected.  +1

**Comment by pkdash** [comment]
@alvacouch I think this PR is ready for merge.

**Pull request merged**